tests/test_serialization.py

```diff
@@ -217,3 +224,93 @@ def test_serialize_fail_llla_different_last_layer_name(laplace, model, model3, r
 
     with pytest.raises(ValueError):
         la2.load_state_dict(torch.load('state_dict.bin'))
+
+
+@pytest.mark.parametrize(
+    "model_device_str",
+    ["cpu", "cuda"] if torch.cuda.is_available() else ["cpu"],
+)
+@pytest.mark.parametrize(
+    "map_location",
+    ["cuda", "cpu"] if torch.cuda.is_available() else ["cpu"],
+)
+@pytest.mark.parametrize("laplace", flavors + flavors_subnet)
+def test_map_location(
+    laplace, model, reg_loader, tmp_path, map_location, model_device_str
+):
+    # Only for LLLaplace, we need dill.
+    #   AttributeError: Can't pickle local object 'FeatureExtractor._get_hook.<locals>.hook'
+    if issubclass(laplace, LLLaplace):
+        if find_spec("dill") is None:
+            pytest.skip(
+                reason="dill package not found but needed for this test"
+            )
+        else:
+            import dill
+
+            torch_save = lambda obj, fn: torch.save(
+                obj, fn, pickle_module=dill
+            )
+    else:
+        # Use default pickle_module=pickle, but no need to import pickle here
+        # just to set pickle_module=pickle.
+        torch_save = lambda obj, fn: torch.save(obj, fn)
+
+    device = torch.device(model_device_str)
+    kwds = dict(model=model.to(device), likelihood="regression")
+    if issubclass(laplace, SubnetLaplace):
+        kwds.update(subnetwork_indices=torch.LongTensor([1, 10, 104, 44]))
+    la = laplace(**kwds)
+
+    # device(type='cuda', index=0) != device(type='cuda')
+    ##assert la._device == device
+    assert la._device.type == device.type
+
+    la.fit(reg_loader)
+    la.optimize_prior_precision()
+    la.sigma_noise = 1231
+
+    save_fn = tmp_path / "la.pt"
+    torch_save(la, save_fn)
+    la2 = torch.load(save_fn, map_location=map_location)
+
+    assert la2._device.type == map_location
+
+    # Test tensor-valued properties that use self._device in their definition.
+    # This is redundant if *all* will be used in the forward call below, where
+    # things would fail if self._device is wrong.
+    for name in [
+        "prior_precision_diag",
+        "prior_mean",
+        "prior_precision",
+        "sigma_noise",
+    ]:
+        assert (
+            getattr(la, name).device.type == device.type
+        ), f"la.{name} failed"
+        assert (
+            getattr(la2, name).device.type == map_location
+        ), f"la2.{name} failed"
+
+    # Test tensor attrs.
+    for name, obj in vars(la).items():
+        # Only LLLaplace instances have a self.X that comes from the train
+        # loader, is assigned in fit() and not explicitly sent to device (it
+        # will be dynamically when used, as is done elsewhere in the code, such
+        # as in BaseLaplace.fit()). self.X won't be affected by self._device.
+        # Since the predict() below passes, we ignore the wrong device here
+        # for now.
+        if issubclass(laplace, LLLaplace) and name == "X":
+            continue
```

Comment:
We ignore the `self.X.device` check  on `LLLaplace` instances because of the reason described in the comment. Feedback from a core dev should clarify if this is OK or how else to handle this.

Reply:
`self.X` is only being saved to infer the last-layer of the `self.model`. If I understand correctly, the `self.model` here is also serialized, right? In this case, we don't need to care about `self.X`, so I'm fine with skipping.

Reply:
Yes, `self.model` is serialized -- this is why `next(self.model.parameters()).device` returns the correct device. In the code, I only see `self.X` used [here](https://github.com/aleximmer/Laplace/blob/2543274aad34e2612d708a58c55272fadbc41626/laplace/lllaplace.py#L109-L113) and [here](https://github.com/aleximmer/Laplace/blob/2543274aad34e2612d708a58c55272fadbc41626/laplace/lllaplace.py#L181-L185) where it is sent to `self._device` if needed.

Reply:
Then let's go forward with ignoring `self.X.device` as you've already done.

Reply:
Agreed.